test/integration/fake_upstream.cc

```diff
@@ -413,6 +414,37 @@ void FakeHttpConnection::onGoAway(Http::GoAwayErrorCode code) {
   ENVOY_LOG(info, "FakeHttpConnection receives GOAWAY: ", code);
 }
 
+void FakeHttpConnection::encodeGoAway() {
+  ASSERT(type_ == Type::HTTP2);
+
+  shared_connection_.connection().dispatcher().post([this]() { codec_->goAway(); });
+}
+
+void FakeHttpConnection::encodeProtocolError() {
+  ASSERT(type_ == Type::HTTP2);
+
+  Http::Http2::ServerConnectionImpl* codec =
+      dynamic_cast<Http::Http2::ServerConnectionImpl*>(codec_.get());
+  if (codec != nullptr) {
+    shared_connection_.connection().dispatcher().post([this, codec]() {
+      expect_protocol_error_ = true;
+      Http::Status status = codec->protocolErrorForTest();
+      ASSERT(Http::getStatusCode(status) == Http::StatusCode::CodecProtocolError);
+    });
+  } else {
```

Comment:
Sorry I think I read the last diff about the brace wrong, apologies. I think this is also kind of strange though. Is the issue here that you need to do the dynamic cast on each type since we have 2 codecs and that is what the early return was before? This is pretty hacky though I guess the legacy code is going away soon so maybe not a huge deal. If you don't do this the tests don't pass on the legacy codec builds?

Reply:
yeah that's right. One option would be to not implement protocolErrorForTest() at all in the legacy codec, and have my test try to detect which codec is used and just skip if it's the legacy codec. I'm not sure how exactly that would work, seems like i'd need `#ifdef` from looking at how the codec is initialized? So not sure if that is any better than what I have

Reply:
Another option would be to use ifdef in this method. That might be easier for someone to find/clean up when the legacy codec is removed?

Reply:
I think it's fine to do this as the code won't compile when the legacy codec is removed. Can you add a TODO though to make it clear?